src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -349,13 +349,19 @@ public function __construct( $html, $use_the_static_create_methods_instead = nul
 
 		$this->state->stack_of_open_elements->set_push_handler(
 			function ( WP_HTML_Token $token ) {
-				$this->element_queue[] = new WP_HTML_Stack_Event( $token, WP_HTML_Stack_Event::PUSH );
+				$is_virtual            = ! isset( $this->state->current_token ) || $this->is_tag_closer();
+				$same_node             = isset( $this->state->current_token ) && $token->node_name === $this->state->current_token->node_name;
+				$provenance            = ( ! $same_node || $is_virtual ) ? 'virtual' : 'real';
```

Comment:
In the push and pop handlers I think we need some explanatory comments.

For virtual nodes, it's not obvious why tag closers are related to virtual nodes. I think this is because when when we find a tag closer but we're in the push handler, it must not be the same node. The opposite is true in the pop handler. We expect to push openers and pop closers.

I'm also curious about the same_node check. Is it sufficient to compare the node name? Is it impossible for the virtual node name to coincide with the current node name?

---

That said, this does seem to work correctly. I added virtual/real information to the debugger (&#x1D456; and &#x211d;), here's what I see:

```html
<a>
<a></a>
<h1><h3></h6>
</p>
```

![Screenshot 2024-06-21 at 21 09 26](https://github.com/WordPress/wordpress-develop/assets/841763/cd070a17-5717-4dcd-946d-a471328ce27c)

(The `P` closer generates a virtual `P` opener at the wrong level, that was an issue before this PR.)

Reply:
Love the addition to the debugger.